sympy/polys/series/python_powerseriesring.py

```diff
@@ -0,0 +1,357 @@
+        return neg_coeffs, prec
+
+    def add(self, s1: USeries[T], s2: USeries[T]) -> USeries[T]:
+        """Add two power series."""
+        return _useries_add(s1, s2, self.domain, self.prec)
+
+    def subtract(self, s1: USeries[T], s2: USeries[T]) -> USeries[T]:
+        """Subtract two power series."""
+        return _useries_sub(s1, s2, self.domain, self.prec)
+
+    def multiply(self, s1: USeries[T], s2: USeries[T]) -> USeries[T]:
+        """Multiply two power series."""
+        return _useries_mul(s1, s2, self.domain, self.prec)
+
+    def multiply_ground(self, s: USeries[T], n: Any) -> USeries[T]:
+        """Multiply a power series by a ground element."""
+        return _useries_mul_ground(s, n, self.domain, self.prec)
+
+    def pow_int(self, s: USeries[T], n: int) -> USeries[T]:
+        """Raise a power series to an integer power."""
+        return _useries_pow_int(s, n, self.domain, self.prec)
+
+    def square(self, s: USeries[T]) -> USeries[T]:
+        """Return the square of a power series."""
+        return _useries_mul(s, s, self.domain, self.prec)
+
+    def truncate(self, s: USeries[T], n: int) -> USeries[T]:
+        """Truncate a power series to the first n terms."""
+        return _useries_truncate(s, n, self.domain)
+
+
+class PythonPowerSeriesRingQQ(PowerSeriesRing[USeries[T]]):
+    """Python implementation of power series ring over rational field."""
+
+    _domain = QQ
+
+    def __init__(self, prec: int = 6) -> None:
+        self._prec = prec
+
+    def __repr__(self) -> str:
+        return f"Python Power Series Ring over {self.domain} with precision {self.prec}"
+
+    def __eq__(self, other: Any) -> bool:
+        if not isinstance(other, PythonPowerSeriesRingQQ):
+            return NotImplemented
+        return self.prec == other.prec
+
+    def __hash__(self) -> int:
+        return hash((self.domain, self.prec))
+
+    @property
+    def domain(self) -> Domain:
+        """Return the ground domain of the power series ring."""
+        return self._domain
+
+    @property
+    def prec(self) -> int:
+        """Return the default precision for power series operations."""
+        return self._prec
+
+    @property
+    def one(self) -> USeries[T]:
+        if self.prec == 0:
+            return ([], 0)
+        return ([QQ(1)], None)
+
+    @property
+    def zero(self) -> USeries[T]:
+        if self.prec == 0:
+            return ([], 0)
+        return ([QQ(0)], None)
+
+    @property
+    def gen(self) -> USeries[T]:
+        if self.prec == 0:
+            return ([], 0)
+        if self.prec == 1:
+            return ([QQ(0)], 1)
+        return ([QQ(1), QQ(0)], None)
+
+    def pretty(self, series: USeries[T]) -> str:
+        coeffs, prec = series
+        return _series_from_list(coeffs[::-1], prec)
+
+    def print(self, series: USeries[T]) -> str:
+        return self.pretty(series)
+
+    def from_list(self, coeffs: DUP[T], prec: int | None = None) -> USeries[T]:
```

Comment:
I would call this `list[T]` rather than `DUP[T]` since this has coefficients in reverse of DUP representation.